swarm_dial.go

```diff
@@ -480,17 +480,37 @@ loop:
 			}
 
 		case <-triggerDial:
-			for _, addr := range nextDial {
+			// we dial batches of addresses together, logically belonging to the same batch
+			// after a batch of addresses has been dialed, we add a delay before initiating the next batch
+			dialed := false
+			last := 0
+			next := 0
+			for i, addr := range nextDial {
+				if dialed && !s.sameAddrBatch(nextDial[last], addr) {
+					break
+				}
+
+				next = i + 1
+
 				// spawn the dial
 				ad := pending[addr]
 				err := s.dialNextAddr(ad.ctx, p, addr, resch)
 				if err != nil {
 					dispatchError(ad, err)
+					continue
 				}
+
+				ad.dialed = true
+				dialed = true
+				last = i
+				active++
 			}
 
-			nextDial = nil
-			triggerDial = nil
+			nextDial = nextDial[next:]
+			if !dialed || len(nextDial) == 0 {
```

Comment:
if `dialed` is false, won't `len(nextDial)` always be 0?

Reply:
yes, but we could have dialed=true and no more addresses, so we need to check nextDial as well.

Reply:
Ah, I see.

Reply:
Ah, wait, no. My point is that we can just get rid of the `dialed` variable, right?

Reply:
I guess we could, but it would be useful when/if we introduce delays so that we know that `last` is valid.